File: piwigo/Image/Image Parameters/Cells/EditImageThumbTableViewCell.m
//
//  EditImageThumbTableViewCell.m
//  piwigo
//
//  Created by Eddy Lelièvre-Berna on 02/01/2020.
//  Copyright © 2020 Piwigo.org. All rights reserved.
//

#import "EditImageThumbTableViewCell.h"
#import "EditImageThumbCollectionViewCell.h"
#import "ImagesCollection.h"

NSString * const kEditImageThumbTableCell_ID = @"EditImageThumbTableCell";

@interface EditImageThumbTableViewCell() <UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout, UIScrollViewDelegate, EditImageThumbnailDelegate>

@property (nonatomic, strong) NSArray<PiwigoImageData *> *images;
@property (nonatomic, strong) IBOutlet UICollectionView *editImageThumbCollectionView;
@property (nonatomic, assign) CGPoint startingScrollingOffset;

@end

@implementation EditImageThumbTableViewCell

-(void)awakeFromNib
{
    [super awakeFromNib];

    // Register thumbnail collection view cell
    [self.editImageThumbCollectionView registerNib:[UINib nibWithNibName:@"EditImageThumbCollectionViewCell" bundle:nil] forCellWithReuseIdentifier:kEditImageThumbCollectionCell_ID];
}

-(void)setupWithImages:(NSArray<PiwigoImageData *> *)imageSelection
{
    // Data
    self.images = imageSelection;
    
    // Collection of images
    self.backgroundColor = [UIColor piwigoCellBackgroundColor];
    if (self.editImageThumbCollectionView == nil) {
        self.editImageThumbCollectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:[UICollectionViewFlowLayout new]];
        [self.editImageThumbCollectionView reloadData];
    } else {
        [self.editImageThumbCollectionView.collectionViewLayout invalidateLayout];
    }
}

-(void)prepareForReuse
{
    [super prepareForReuse];
}


#pragma mark - UICollectionViewDataSource Methods

-(NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView
{
    return 1;
}

-(NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    // Returns number of images or albums
    return self.images.count;
}

-(UICollectionViewCell*)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    EditImageThumbCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:kEditImageThumbCollectionCell_ID forIndexPath:indexPath];
    [cell setupWithImage:self.images[indexPath.row] removeOption:(self.images.count > 1)];
    cell.delegate = self;
    return cell;
}


#pragma mark - UICollectionViewDelegateFlowLayout Methods

-(UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetForSectionAtIndex:(NSInteger)section
{
    // Avoid unwanted spaces
    if (@available(iOS 13.0, *)) {
        return UIEdgeInsetsMake(0, kImageDetailsMarginsSpacing, 0, kImageDetailsMarginsSpacing);
    } else {
        return UIEdgeInsetsMake(10, kImageDetailsMarginsSpacing, 0, kImageDetailsMarginsSpacing);
    }
}

-(CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section;
{
    return (CGFloat)kImageDetailsCellSpacing;
}

-(CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    return CGSizeMake([ImagesCollection imageDetailsSizeForView:self], 144);
}


#pragma mark - EditImageThumbnailDelegate Methods

-(void)didDeselectImageWithId:(NSInteger)imageId
{
    // Update data source
    NSMutableArray *newImages = [[NSMutableArray alloc] initWithArray:self.images];
    for (PiwigoImageData *imageData in self.images)
    {
        if (imageData.imageId == imageId)
        {
            [newImages removeObject:imageData];
        }
    }
    self.images = newImages;
    [self.editImageThumbCollectionView reloadData];

    // Deselect image in parent view
    if ([self.delegate respondsToSelector:@selector(didDeselectImageWithId:)])
    {
        [self.delegate didDeselectImageWithId:imageId];
    }
}

-(void)didRenameFileOfImageWithId:(NSInteger)imageId andFilename:(NSString *)fileName
{
    // Update data source
    PiwigoImageData *updatedImage;
    NSMutableArray *updatedImages = [[NSMutableArray alloc] init];
    for (PiwigoImageData *imageData in self.images)
    {
        if (imageData.imageId == imageId)
        {
            if (fileName) imageData.fileName = fileName;
            updatedImage = imageData;
        }
        [updatedImages addObject:imageData];
    }
    self.images = updatedImages;
    
    // Update parent image view
    if ([self.delegate respondsToSelector:@selector(didRenameFileOfImage:)])
    {
        [self.delegate didRenameFileOfImage:updatedImage];
    }
}


#pragma mark - ScrollView Manager
- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
    self.startingScrollingOffset = scrollView.contentOffset;
}

-(void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset
{
    CGFloat cellWidth = [self collectionView:self.editImageThumbCollectionView layout:self.editImageThumbCollectionView.collectionViewLayout sizeForItemAtIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]].width + kImageDetailsMarginsSpacing/2.0;
    double offset = scrollView.contentOffset.x + scrollView.contentInset.left;
    double proposedPage = offset / fmax(1.0, cellWidth);
    CGFloat snapPoint = 0.1;
    CGFloat snapDelta = offset > self.startingScrollingOffset.x ? (1 - snapPoint) : snapPoint;

    double page;
    if (floor(proposedPage + snapDelta) == floor(proposedPage)) {
        page = floor(proposedPage);
    }
    else {
        page = floor(proposedPage + 1);
    }

    *targetContentOffset = CGPointMake(cellWidth * page, (*targetContentOffset).y);
}

@end

File: piwigo/Image/Parameters/Cells/EditImageThumbTableViewCell.swift
//
//  EditImageThumbTableViewCell.swift
//  piwigo
//
//  Created by Eddy Lelièvre-Berna on 02/01/2020.
//  Copyright © 2020 Piwigo.org. All rights reserved.
//
//  Converted to Swift 5.3 by Eddy Lelièvre-Berna on 30/08/2021.
//

import UIKit
import piwigoKit

@objc protocol EditImageThumbnailCellDelegate: NSObjectProtocol {
    func didDeselectImage(withID imageID: Int64)
    func didRenameFileOfImage(_ imageData: Image)
}

class EditImageThumbTableViewCell: UITableViewCell, UICollectionViewDelegate
{
    weak var delegate: (any EditImageThumbnailCellDelegate)?
    
    @IBOutlet private var editImageThumbCollectionView: UICollectionView!

    var user: User!
    private var images: [Image]?
    private var startingScrollingOffset = CGPoint.zero

    override func awakeFromNib() {
        super.awakeFromNib()

        // Register thumbnail collection view cell
        editImageThumbCollectionView.register(UINib(nibName: "EditImageThumbCollectionViewCell",
            bundle: nil), forCellWithReuseIdentifier: "EditImageThumbCollectionViewCell")
    }

    func config(withImages imageSelection: [Image]?) {
        // Data
        images = imageSelection

        // Collection of images
        backgroundColor = PwgColor.cellBackground
        if editImageThumbCollectionView == nil {
            editImageThumbCollectionView = UICollectionView(frame: CGRect.zero,
                                                            collectionViewLayout: UICollectionViewFlowLayout())
            editImageThumbCollectionView.reloadData()
        } else {
            editImageThumbCollectionView.collectionViewLayout.invalidateLayout()
        }
    }
}

    
// MARK: - UICollectionViewDataSource Methods
extension EditImageThumbTableViewCell: UICollectionViewDataSource
{
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 1
    }

    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        // Returns number of images or albums
        return images?.count ?? 0
    }

    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "EditImageThumbCollectionViewCell", for: indexPath) as? EditImageThumbCollectionViewCell else {
            debugPrint("Error: collectionView.dequeueReusableCell does not return a EditImageThumbCollectionViewCell!")
            return EditImageThumbCollectionViewCell()
        }
        cell.config(withImage: images?[indexPath.row], removeOption: ((images?.count ?? 0) > 1))
        cell.delegate = self
        return cell
    }
}


// MARK: - UICollectionViewDelegateFlowLayout Methods
extension EditImageThumbTableViewCell: UICollectionViewDelegateFlowLayout
{
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, insetForSectionAt section: Int) -> UIEdgeInsets {
        // Avoid unwanted spaces
        return UIEdgeInsets(top: 0, left: AlbumUtilities.kImageDetailsMarginsSpacing,
                            bottom: 0, right: AlbumUtilities.kImageDetailsMarginsSpacing)
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return AlbumUtilities.kImageDetailsCellSpacing
    }

    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: AlbumUtilities.imageDetailsSize(forView: self), height: 152.0)
    }
}


// MARK: - EditImageThumbnailDelegate Methods
extension EditImageThumbTableViewCell: EditImageThumbnailDelegate
{
    @objc func didDeselectImage(withID imageID: Int64) {
        // Update data source
        let newImages = images?.filter({ $0.pwgID != imageID })
        images = newImages
        editImageThumbCollectionView.reloadData()

        // Deselect image in parent view
        delegate?.didDeselectImage(withID: imageID)
    }

    @objc func didRenameFileOfImage(withId imageID: Int64, andFilename fileName: String) {
        // Retrieve image data from cache
        guard let imageToUpdate = images?.first(where: {$0.pwgID == imageID}) else { return }
        
        // Update image in cache
        imageToUpdate.fileName = fileName

        // Update parent image view
        delegate?.didRenameFileOfImage(imageToUpdate)
    }
}


// MARK: - UIScrollViewDelegate Methods
extension EditImageThumbTableViewCell: UIScrollViewDelegate
{
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        startingScrollingOffset = scrollView.contentOffset
    }

    func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>)
    {
        let cellWidth: CGFloat = collectionView(editImageThumbCollectionView, layout: editImageThumbCollectionView.collectionViewLayout, sizeForItemAt: IndexPath(row: 0, section: 0)).width + AlbumUtilities.kImageDetailsMarginsSpacing / 2.0
        let offset:CGFloat = scrollView.contentOffset.x + scrollView.contentInset.left
        let proposedPage: CGFloat = offset / fmax(1.0, cellWidth)
        let floorProposedPage: CGFloat = floor(proposedPage)
        let snapPoint: CGFloat = 0.1
        let snapDelta: CGFloat = offset > startingScrollingOffset.x ? (1 - snapPoint) : snapPoint

        var page: CGFloat
        if floor(proposedPage + snapDelta) == floorProposedPage {
            page = floorProposedPage
        } else {
            page = floor(proposedPage + 1)
        }

        targetContentOffset.pointee = CGPoint(x: cellWidth * page,
                                              y: targetContentOffset.pointee.y)
    }
}
